runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/jobsubmission/JobService.java

```diff
@@ -0,0 +1,340 @@
+      Struct pipelineOptions = request.getPipelineOptions();
+      if (pipelineOptions == null) {
+        LOG.trace("PIPELINE OPTIONS IS NULL");
+        throw new NullPointerException("Encountered null pipeline options.");
+            /*
+        LOG.debug("Encountered null pipeline options.  Using default.");
+        pipelineOptions = Struct.getDefaultInstance();
+        */
+      } else {
+        LOG.trace("PIPELINE OPTIONS IS NOT NULL");
+      }
+      LOG.trace("PIPELINE OPTIONS {} {}", pipelineOptions.getClass(), pipelineOptions);
+      JobPreparation preparation =
+          JobPreparation
+              .builder()
+              .setId(preparationId)
+              .setPipeline(request.getPipeline())
+              .setOptions(pipelineOptions)
+              .setStagingService(stagingService)
+              .build();
+      JobPreparation previous = preparations.putIfAbsent(preparationId, preparation);
+      if (previous != null) {
+        // retry recursively in the unlikely case of a name collision.
+        String errMessage =
+            String.format("Name collision for preparation ID \"%s\". Retrying.", preparationId);
+        LOG.warn(errMessage);
+        prepare(request, responseObserver);
+        return;
+      }
+
+      // send response
+      PrepareJobResponse response =
+          PrepareJobResponse
+              .newBuilder()
+              .setPreparationId(preparationId)
+              .setArtifactStagingEndpoint(stagingService.getApiServiceDescriptor())
+              .build();
+      responseObserver.onNext(response);
+      responseObserver.onCompleted();
+    } catch (Exception e) {
+      LOG.error("Could not prepare job with name {}", request.getJobName(), e);
+      responseObserver.onError(Status.INTERNAL.withCause(e).asException());
+    }
+  }
+
+  @Override
+  public void run(
+      RunJobRequest request, StreamObserver<RunJobResponse> responseObserver) {
+    LOG.trace("{} {}", RunJobRequest.class.getSimpleName(), request);
+
+    String preparationId = request.getPreparationId();
+    try {
+      // retrieve job preparation
+      JobPreparation preparation = preparations.get(preparationId);
+      if (preparation == null) {
+        String errMessage = String.format("Unknown Preparation Id \"%s\".", preparationId);
+        StatusException exception = Status.NOT_FOUND.withDescription(errMessage).asException();
+        responseObserver.onError(exception);
+        return;
+      }
+
+      // create new invocation
+      JobInvocation invocation = invoker.invoke(preparation, request.getStagingToken());
+      String invocationId = invocation.getId();
+      invocation.start();
+      invocations.put(invocationId, invocation);
+      RunJobResponse response =
+          RunJobResponse.newBuilder().setJobId(invocationId).build();
+      responseObserver.onNext(response);
+      responseObserver.onCompleted();
+    } catch (StatusRuntimeException e) {
```

Comment:
Not sure why the handling of this exception differs from the next one, maybe we can join them?

Reply:
In general if we encounter an exception, it should be treated as an internal exception and wrapped in a StatusRuntimeException that reflects this.  However if the JobInvoker throws a StatusRuntimeException during the process of its invocation, the status code might be something other than INTERNAL.  In this case we prefer to preserve that code.